File: agent-python/agent.py
#!/usr/bin/env python3
"""
Spark Lab Agent - Python 版本
轻量级服务器代理，支持 Python 3.6+
"""

import json
import os
import sys
import time
import subprocess
import socketio
import docker
import psutil
from threading import Thread

class SparkLabAgent:
    def __init__(self, config_path):
        self.config = self.load_config(config_path)
        self.docker_client = docker.from_env()
        self.sio = socketio.Client(reconnection=True, reconnection_delay=5)
        self.setup_handlers()
        
    def load_config(self, config_path):
        with open(config_path, 'r') as f:
            return json.load(f)
    
    def setup_handlers(self):
        @self.sio.event
        def connect():
            print('[Agent] 已连接到中心服务器')
            self.send_heartbeat()
            
        @self.sio.event
        def disconnect():
            print('[Agent] 断开连接')
            
        @self.sio.on('docker:create')
        def on_create(data):
            print(f'[Agent] 收到 docker:create 请求')
            result = self.handle_create(data)
            self.sio.emit('docker:create:response', result, namespace='/agent')
            
        @self.sio.on('docker:start')
        def on_start(data):
            print(f'[Agent] 收到 docker:start 请求')
            result = self.handle_start(data)
            self.sio.emit('docker:start:response', result, namespace='/agent')
            
        @self.sio.on('docker:stop')
        def on_stop(data):
            print(f'[Agent] 收到 docker:stop 请求')
            result = self.handle_stop(data)
            self.sio.emit('docker:stop:response', result, namespace='/agent')
            
        @self.sio.on('docker:remove')
        def on_remove(data):
            print(f'[Agent] 收到 docker:remove 请求')
            result = self.handle_remove(data)
            self.sio.emit('docker:remove:response', result, namespace='/agent')
            
        @self.sio.on('docker:exec')
        def on_exec(data):
            print(f'[Agent] 收到 docker:exec 请求')
            result = self.handle_exec(data)
            self.sio.emit('docker:exec:response', result, namespace='/agent')
        
        @self.sio.on('docker:list')
        def on_list(data):
            print(f'[Agent] 收到 docker:list 请求, 数据: {data}')
            result = self.handle_list(data)
            containers_count = len(result.get("data", {}).get("containers", []) if result.get("success") else [])
            print(f'[Agent] 返回容器列表: {containers_count} 个容器')
            print(f'[Agent] 发送响应事件: docker:list:response')
            self.sio.emit('docker:list:response', result, namespace='/agent')
            
        @self.sio.on('docker:run')
        def on_run(data):
            print(f'[Agent] 收到 docker:run 请求')
            result = self.handle_run(data)
            self.sio.emit('docker:run:response', result, namespace='/agent')
            
        @self.sio.on('docker:exec_create')
        def on_exec_create(data):
            print(f'[Agent] 收到 docker:exec_create 请求')
            result = self.handle_exec_create(data)
            self.sio.emit('docker:exec_create:response', result, namespace='/agent')
            
        @self.sio.on('docker:exec_start')
        def on_exec_start(data):
            print(f'[Agent] 收到 docker:exec_start 请求')
            result = self.handle_exec_start(data)
            self.sio.emit('docker:exec_start:response', result, namespace='/agent')
    
    def connect(self):
        print(f"[Agent] 连接到: {self.config['serverUrl']}")
        self.sio.connect(
            self.config['serverUrl'],
            namespaces=['/agent'],
            auth={
                'token': self.config['agentToken'],
                'serverName': self.config['serverName']
            }
        )
        
        # 启动心跳线程
        Thread(target=self.heartbeat_loop, daemon=True).start()
        
    def heartbeat_loop(self):
        # 等待连接建立
        time.sleep(1)
        # 立即发送第一次心跳
        if self.sio.connected:
            self.send_heartbeat()
        # 然后每 0.5 秒发送一次
        while True:
            time.sleep(0.5)
            if self.sio.connected:
                self.send_heartbeat()
    
    def send_heartbeat(self):
        try:
            cpu_percent = psutil.cpu_percent(interval=1)
            mem = psutil.virtual_memory()
            containers = self.docker_client.containers.list()
            
            # 获取 CPU 信息
            cpu_count = psutil.cpu_count(logical=False) or psutil.cpu_count()
            cpu_model = 'Unknown'
            try:
                with open('/proc/cpuinfo', 'r') as f:
                    for line in f:
                        if 'model name' in line:
                            cpu_model = line.split(':')[1].strip()
                            break
            except:
                pass
            
            # 获取本机 IP 地址
            server_ip = 'unknown'
            try:
                import socket
                s = socket.socket(socket.AF_INET, socket.SOCK_DGRAM)
                s.connect(("8.8.8.8", 80))
                server_ip = s.getsockname()[0]
                s.close()
            except:
                pass
            
            stats = {
                'cpuUsage': cpu_percent,
                'memoryUsage': mem.percent,
                'totalMemory': mem.total // (1024 * 1024),
                'cpuCores': cpu_count,
                'cpuModel': cpu_model,
                'activeContainers': len(containers),
                'serverIp': server_ip,
                'timestamp': time.strftime('%Y-%m-%dT%H:%M:%S')
            }
            
            print(f'[Agent] 发送心跳: CPU={cpu_percent:.1f}%, 内存={mem.percent:.1f}%, 核心={cpu_count}, 总内存={mem.total // (1024 * 1024)}MB, IP={server_ip}')
            self.sio.emit('agent:heartbeat', stats, namespace='/agent')
        except Exception as e:
            print(f'[Agent] 心跳失败: {e}')
    
    def handle_create(self, data):
        try:
            print(f"[Agent] 创建容器: {data['name']}")
            
            # 拉取镜像
            try:
                self.docker_client.images.get(data['image'])
            except docker.errors.ImageNotFound:
                print(f"[Agent] 拉取镜像: {data['image']}")
                self.docker_client.images.pull(data['image'])
            
            # 创建容器
            container = self.docker_client.containers.run(
                data['image'],
                name=data['name'],
                detach=True,
                tty=True,
                stdin_open=True,
                mem_limit=f"{data['memoryLimit']}m",
                nano_cpus=int(data['cpuLimit'] * 1e9),
                ports={
                    '22/tcp': None,
                    '5900/tcp': None,
                    '8080/tcp': None
                },
                publish_all_ports=True
            )
            
            # 获取端口映射
            container.reload()
            ports = container.attrs['NetworkSettings']['Ports']
            
            return {
                'success': True,
                'data': {
                    'id': container.id,
                    'sshPort': int(ports.get('22/tcp', [{}])[0].get('HostPort', 0)) or None,
                    'vncPort': int(ports.get('5900/tcp', [{}])[0].get('HostPort', 0)) or None,
                    'idePort': int(ports.get('8080/tcp', [{}])[0].get('HostPort', 0)) or None
                }
            }
        except Exception as e:
            print(f'[Agent] 创建失败: {e}')
            return {'success': False, 'error': str(e)}
    
    def handle_start(self, data):
        try:
            container = self.docker_client.containers.get(data['containerId'])
            container.start()
            return {'success': True}
        except Exception as e:
            return {'success': False, 'error': str(e)}
    
    def handle_stop(self, data):
        try:
            container = self.docker_client.containers.get(data['containerId'])
            container.stop()
            return {'success': True}
        except Exception as e:
            return {'success': False, 'error': str(e)}
    
    def handle_remove(self, data):
        try:
            container = self.docker_client.containers.get(data['containerId'])
            container.remove(force=True)
            return {'success': True}
        except Exception as e:
            return {'success': False, 'error': str(e)}
    
    def handle_exec(self, data):
        try:
            container = self.docker_client.containers.get(data['containerId'])
            cmd = '/bin/sh -c "' + data['command'] + '"'
            result = container.exec_run(cmd, stdout=True, stderr=True)
            return {
                'success': True,
                'output': result.output.decode('utf-8', errors='ignore')
            }
        except Exception as e:
            return {'success': False, 'error': str(e)}
    
    def handle_list(self, data):
        try:
            # 获取所有容器（包括停止的）
            all_containers = self.docker_client.containers.list(all=True)
            containers_info = []
            
            for container in all_containers:
                containers_info.append({
                    'id': container.id,
                    'name': container.name,
                    'image': container.image.tags[0] if container.image.tags else container.image.id[:12],
                    'status': container.status,
                    'created': container.attrs['Created'],
                    'ports': container.attrs['NetworkSettings']['Ports'],
                })
            
            return {'success': True, 'data': {'containers': containers_info}}
        except Exception as e:
            return {'success': False, 'error': str(e)}
    
    def handle_run(self, data):
        try:
            print(f"[Agent] 使用完整 docker run 命令创建容器")
            import shlex
            
            # 解析命令
            command = data.get('command', '')
            print(f"[Agent] 执行命令: {command}")
            
            # 使用 subprocess 执行命令
            import subprocess
            result = subprocess.run(
                command,
                shell=True,
                capture_output=True,
                text=True
            )
            
            if result.returncode != 0:
                print(f"[Agent] 命令执行失败: {result.stderr}")
                return {'success': False, 'error': result.stderr}
            
            print(f"[Agent] 命令执行成功: {result.stdout}")
            
            # 获取最新创建的容器
            containers = self.docker_client.containers.list(all=True)
            if not containers:
                return {'success': False, 'error': '无法找到创建的容器'}
            
            # 按创建时间排序，取最新的
            containers.sort(key=lambda c: c.attrs['Created'], reverse=True)
            container = containers[0]
            container.reload()
            ports = container.attrs['NetworkSettings']['Ports']
            
            return {
                'success': True,
                'data': {
                    'id': container.id,
                    'sshPort': int(ports.get('22/tcp', [{}])[0].get('HostPort', 0)) or None,
                    'vncPort': int(ports.get('5900/tcp', [{}])[0].get('HostPort', 0)) or None,
                    'idePort': int(ports.get('8080/tcp', [{}])[0].get('HostPort', 0)) or None
                }
            }
        except Exception as e:
            print(f'[Agent] docker:run 失败: {e}')
            return {'success': False, 'error': str(e)}
    
    def handle_create(self, data):
        try:
            print(f"[Agent] 创建容器: {data['name']}")
            
            # 拉取镜像
            try:
                self.docker_client.images.get(data['image'])
            except docker.errors.ImageNotFound:
                print(f"[Agent] 拉取镜像: {data['image']}")
                self.docker_client.images.pull(data['image'])
            
            # 构建容器参数
            run_kwargs = {
                'image': data['image'],
                'name': data['name'],
                'detach': True,
                'tty': True,
                'stdin_open': True,
                'mem_limit': f"{data['memoryLimit']}m",
                'nano_cpus': int(data['cpuLimit'] * 1e9),
            }
            
            # 添加环境变量
            if 'environment' in data and data['environment']:
                import json
                env_dict = json.loads(data['environment'])
                run_kwargs['environment'] = [f"{k}={v}" for k, v in env_dict.items()]
            
            # 添加端口映射
            ports_config = {
                '22/tcp': None,
                '5900/tcp': None,
                '8080/tcp': None
            }
            if 'ports' in data and data['ports']:
                import json
                custom_ports = json.loads(data['ports'])
                ports_config.update(custom_ports)
            run_kwargs['ports'] = ports_config
            
            # 添加卷挂载
            if 'volumes' in data and data['volumes']:
                import json
                run_kwargs['volumes'] = json.loads(data['volumes'])
            
            # 添加重启策略
            if 'restartPolicy' in data and data['restartPolicy']:
                import json
                run_kwargs['restart_policy'] = json.loads(data['restartPolicy'])
            
            # 添加网络模式
            if 'networkMode' in data and data['networkMode']:
                run_kwargs['network_mode'] = data['networkMode']
            
            # 添加主机名
            if 'hostname' in data and data['hostname']:
                run_kwargs['hostname'] = data['hostname']
            
            # 添加工作目录
            if 'workingDir' in data and data['workingDir']:
                run_kwargs['working_dir'] = data['workingDir']
            
            # 添加标签
            if 'labels' in data and data['labels']:
                import json
                run_kwargs['labels'] = json.loads(data['labels'])
            
            # 添加启动命令
            if 'startupCommand' in data and data['startupCommand']:
                run_kwargs['command'] = data['startupCommand']
            
            # 创建并启动容器
            container = self.docker_client.containers.run(**run_kwargs)
            
            # 获取端口映射
            container.reload()
            ports = container.attrs['NetworkSettings']['Ports']
            
            return {
                'success': True,
                'data': {
                    'id': container.id,
                    'sshPort': int(ports.get('22/tcp', [{}])[0].get('HostPort', 0)) or None,
                    'vncPort': int(ports.get('5900/tcp', [{}])[0].get('HostPort', 0)) or None,
                    'idePort': int(ports.get('8080/tcp', [{}])[0].get('HostPort', 0)) or None
                }
            }
        except Exception as e:
            print(f'[Agent] 创建失败: {e}')
            return {'success': False, 'error': str(e)}
    
    def handle_exec_create(self, data):
        try:
            print(f"[Agent] 创建 exec 实例: {data.get('command')}")
            container = self.docker_client.containers.get(data['containerId'])
            
            # 创建 exec 实例
            exec_instance = container.client.api.exec_create(
                container.id,
                cmd=data.get('command', '/bin/bash'),
                tty=data.get('tty', True),
                stdin=data.get('stdin', True),
                stdout=data.get('stdout', True),
                stderr=data.get('stderr', True)
            )
            
            return {
                'success': True,
                'data': {
                    'execId': exec_instance['Id']
                }
            }
        except Exception as e:
            print(f'[Agent] exec_create 失败: {e}')
            return {'success': False, 'error': str(e)}
    
    def handle_exec_start(self, data):
        try:
            print(f"[Agent] 启动 exec 实例: {data.get('execId')}")
            
            # 启动 exec 实例
            output = self.docker_client.api.exec_start(
                data['execId'],
                stream=data.get('stream', False),
                detach=data.get('detach', False),
                tty=data.get('tty', True)
            )
            
            if data.get('stream', False):
                # 流式输出，返回标识让前端处理
                return {
                    'success': True,
                    'data': {
                        'streaming': True
                    }
                }
            else:
                # 非流式输出，直接返回结果
                return {
                    'success': True,
                    'data': {
                        'output': output.decode('utf-8', errors='ignore')
                    }
                }
        except Exception as e:
            print(f'[Agent] exec_start 失败: {e}')
            return {'success': False, 'error': str(e)}
    
    def run(self):
        self.connect()
        print('[Agent] 已启动，按 Ctrl+C 退出')
        self.sio.wait()

if __name__ == '__main__':
    config_path = os.getenv('CONFIG_PATH', '/etc/spark-lab-agent/config.json')
    
    if not os.path.exists(config_path):
        print(f'错误: 配置文件不存在: {config_path}')
        sys.exit(1)
    
    agent = SparkLabAgent(config_path)
    
    try:
        agent.run()
    except KeyboardInterrupt:
        print('\n[Agent] 正在关闭...')
        sys.exit(0)

File: backend-go/internal/handler/agent.go
package handler

import (
	"fmt"
	"net/http"
	"os"
	"strings"
	"time"

	"bigdata_zhoc/backend-go/internal/model"

	"github.com/gin-gonic/gin"
)

type agentAuthReq struct {
	Token      string  `json:"token"`
	ServerName string  `json:"serverName"`
	ServerIP   *string `json:"serverIp"`
	CPUCores   *int    `json:"cpuCores"`
	TotalMemory *int   `json:"totalMemory"`
	CPUModel   *string `json:"cpuModel"`
}

type agentHeartbeatReq struct {
	Token            string   `json:"token"`
	ServerName       string   `json:"serverName"`
	CPUUsage         *float64 `json:"cpuUsage"`
	MemoryUsage      *float64 `json:"memoryUsage"`
	ActiveContainers *int     `json:"activeContainers"`
	ServerIP         *string  `json:"serverIp"`
	CPUCores         *int     `json:"cpuCores"`
	TotalMemory      *int     `json:"totalMemory"`
	CPUModel         *string  `json:"cpuModel"`
}

func (h *Handler) findServerByAgentToken(serverName, token string) (*model.Server, bool) {
	var server model.Server
	err := h.db.Where("name = ? AND password = ?", serverName, token).First(&server).Error
	if err != nil {
		return nil, false
	}
	return &server, true
}

func normalizeAgentIP(ip *string) *string {
	if ip == nil {
		return nil
	}
	v := strings.TrimSpace(*ip)
	if v == "" || strings.EqualFold(v, "unknown") {
		return nil
	}
	return &v
}

func (h *Handler) AgentRegister(c *gin.Context) {
	var req agentAuthReq
	if err := c.ShouldBindJSON(&req); err != nil || strings.TrimSpace(req.Token) == "" || strings.TrimSpace(req.ServerName) == "" {
		c.JSON(http.StatusBadRequest, gin.H{"message": "token and serverName are required"})
		return
	}

	server, ok := h.findServerByAgentToken(req.ServerName, req.Token)
	if !ok {
		c.JSON(http.StatusUnauthorized, gin.H{"message": "Invalid token or server name"})
		return
	}

	updates := map[string]any{
		"status":      "online",
		"lastCheckAt": time.Now(),
		"updatedAt":   time.Now(),
	}
	if ip := normalizeAgentIP(req.ServerIP); ip != nil {
		updates["host"] = *ip
	}
	if req.CPUCores != nil && *req.CPUCores > 0 {
		updates["cpuCores"] = *req.CPUCores
	}
	if req.TotalMemory != nil && *req.TotalMemory > 0 {
		updates["totalMemory"] = *req.TotalMemory
	}
	if req.CPUModel != nil && strings.TrimSpace(*req.CPUModel) != "" {
		updates["cpuModel"] = strings.TrimSpace(*req.CPUModel)
	}

	if err := h.db.Model(&model.Server{}).Where("id = ?", server.ID).Updates(updates).Error; err != nil {
		c.JSON(http.StatusInternalServerError, gin.H{"message": "Register agent failed"})
		return
	}

	c.JSON(http.StatusOK, gin.H{"message": "Agent registered", "serverId": server.ID})
}

func (h *Handler) AgentHeartbeat(c *gin.Context) {
	var req agentHeartbeatReq
	if err := c.ShouldBindJSON(&req); err != nil || strings.TrimSpace(req.Token) == "" || strings.TrimSpace(req.ServerName) == "" {
		c.JSON(http.StatusBadRequest, gin.H{"message": "token and serverName are required"})
		return
	}

	server, ok := h.findServerByAgentToken(req.ServerName, req.Token)
	if !ok {
		c.JSON(http.StatusUnauthorized, gin.H{"message": "Invalid token or server name"})
		return
	}

	updates := map[string]any{
		"status":      "online",
		"lastCheckAt": time.Now(),
		"updatedAt":   time.Now(),
	}
	if req.CPUUsage != nil {
		updates["cpuUsage"] = *req.CPUUsage
	}
	if req.MemoryUsage != nil {
		updates["memoryUsage"] = *req.MemoryUsage
	}
	if req.ActiveContainers != nil {
		updates["activeContainers"] = *req.ActiveContainers
	}
	if req.CPUCores != nil && *req.CPUCores > 0 {
		updates["cpuCores"] = *req.CPUCores
	}
	if req.TotalMemory != nil && *req.TotalMemory > 0 {
		updates["totalMemory"] = *req.TotalMemory
	}
	if req.CPUModel != nil && strings.TrimSpace(*req.CPUModel) != "" {
		updates["cpuModel"] = strings.TrimSpace(*req.CPUModel)
	}
	if ip := normalizeAgentIP(req.ServerIP); ip != nil {
		updates["host"] = *ip
	}

	if err := h.db.Model(&model.Server{}).Where("id = ?", server.ID).Updates(updates).Error; err != nil {
		c.JSON(http.StatusInternalServerError, gin.H{"message": "Heartbeat update failed"})
		return
	}

	c.JSON(http.StatusOK, gin.H{"ok": true})
}

func (h *Handler) AgentInstallScript(c *gin.Context) {
	script := `#!/bin/bash

set -e

echo "================================"
echo "Spark Lab Agent 安装程序 (Python)"
echo "================================"

if [ $# -lt 3 ]; then
    echo "错误: 缺少参数"
    echo "用法: bash install.sh <SERVER_URL> <TOKEN> <SERVER_NAME>"
    exit 1
fi

SERVER_URL=$1
AGENT_TOKEN=$2
SERVER_NAME=$3
MAX_CONTAINERS=${4:-10}

echo "服务器地址: $SERVER_URL"
echo "服务器名称: $SERVER_NAME"

if [ "$EUID" -ne 0 ]; then
    echo "错误: 请使用 root 权限运行"
    exit 1
fi

if ! command -v docker &> /dev/null; then
    echo "错误: Docker 未安装"
    echo "安装: curl -fsSL https://get.docker.com | sh"
    exit 1
fi

if ! command -v python3 &> /dev/null; then
    echo "正在安装 Python 3..."
    if [ -f /etc/debian_version ]; then
        apt-get update
        apt-get install -y python3 python3-pip python3-dev gcc
    elif [ -f /etc/redhat-release ]; then
        yum install -y python3 python3-pip python3-devel gcc
    fi
fi

INSTALL_DIR="/opt/spark-lab-agent"
CONFIG_DIR="/etc/spark-lab-agent"
mkdir -p $INSTALL_DIR
mkdir -p $CONFIG_DIR

echo "下载 Agent 程序..."
curl -fsSL "$SERVER_URL/agent/agent-http.py" -o "$INSTALL_DIR/agent.py"
chmod +x "$INSTALL_DIR/agent.py"

echo "安装 Python 依赖..."
pip3 install python-socketio[client]==5.7.0 docker==5.0.3 psutil==5.9.0 -i https://pypi.tuna.tsinghua.edu.cn/simple

cat > "$CONFIG_DIR/config.json" <<EOF
{
  "serverUrl": "$SERVER_URL",
  "agentToken": "$AGENT_TOKEN",
  "serverName": "$SERVER_NAME",
  "maxContainers": $MAX_CONTAINERS
}
EOF

chmod 600 "$CONFIG_DIR/config.json"

echo "配置 Docker TCP 端口..."
cat > /etc/docker/daemon.json <<EOF
{
  "hosts": ["unix:///var/run/docker.sock", "tcp://0.0.0.0:2375"]
}
EOF

mkdir -p /etc/systemd/system/docker.service.d
cat > /etc/systemd/system/docker.service.d/override.conf <<EOF
[Service]
ExecStart=
ExecStart=/usr/bin/dockerd
EOF

systemctl daemon-reload
systemctl restart docker

cat > /etc/systemd/system/spark-lab-agent.service <<EOF
[Unit]
Description=Spark Lab Agent (Python)
After=network.target docker.service
Requires=docker.service

[Service]
Type=simple
User=root
WorkingDirectory=$INSTALL_DIR
Environment="CONFIG_PATH=$CONFIG_DIR/config.json"
ExecStart=/usr/bin/python3 $INSTALL_DIR/agent.py
Restart=always
RestartSec=10
StandardOutput=journal
StandardError=journal

[Install]
WantedBy=multi-user.target
EOF

systemctl daemon-reload
systemctl enable spark-lab-agent
systemctl start spark-lab-agent

echo ""
echo "================================"
echo "✓ 安装完成！"
echo "================================"
echo "查看日志: journalctl -u spark-lab-agent -f"
`

	c.Header("Content-Type", "text/plain; charset=utf-8")
	c.String(http.StatusOK, script)
}

func (h *Handler) AgentHTTPPythonFile(c *gin.Context) {
	script := `#!/usr/bin/env python3
import json
import os
import sys
import time
import requests

try:
	import psutil
except Exception:
	psutil = None

try:
	import docker
except Exception:
	docker = None


class SparkLabHTTPAgent:
	def __init__(self, config_path):
		with open(config_path, 'r', encoding='utf-8') as f:
			self.config = json.load(f)
		self.server_url = self.config['serverUrl'].rstrip('/')
		self.token = self.config['agentToken']
		self.server_name = self.config['serverName']
		self.session = requests.Session()
		self.docker_client = None
		if docker is not None:
			try:
				self.docker_client = docker.from_env()
			except Exception:
				self.docker_client = None

	def get_ip(self):
		try:
			import socket
			s = socket.socket(socket.AF_INET, socket.SOCK_DGRAM)
			s.connect(("8.8.8.8", 80))
			ip = s.getsockname()[0]
			s.close()
			return ip
		except Exception:
			return None

	def collect_stats(self):
		cpu_usage = 0.0
		memory_usage = 0.0
		total_memory = 0
		cpu_cores = 0
		cpu_model = "Unknown"
		active_containers = 0

		if psutil is not None:
			try:
				cpu_usage = float(psutil.cpu_percent(interval=0.5))
			except Exception:
				pass
			try:
				memory_usage = float(psutil.virtual_memory().percent)
				total_memory = int(psutil.virtual_memory().total // (1024 * 1024))
			except Exception:
				pass
			try:
				cpu_cores = int(psutil.cpu_count(logical=False) or psutil.cpu_count() or 0)
			except Exception:
				pass

		try:
			with open('/proc/cpuinfo', 'r', encoding='utf-8', errors='ignore') as f:
				for line in f:
					if 'model name' in line:
						cpu_model = line.split(':', 1)[1].strip()
						break
		except Exception:
			pass

		if self.docker_client is not None:
			try:
				active_containers = len(self.docker_client.containers.list())
			except Exception:
				pass

		return {
			'cpuUsage': cpu_usage,
			'memoryUsage': memory_usage,
			'totalMemory': total_memory,
			'cpuCores': cpu_cores,
			'cpuModel': cpu_model,
			'activeContainers': active_containers,
			'serverIp': self.get_ip(),
		}

	def post_json(self, path, payload):
		url = self.server_url + path
		r = self.session.post(url, json=payload, timeout=10)
		if r.status_code >= 400:
			raise RuntimeError(f"HTTP {r.status_code}: {r.text}")
		return r.json() if r.text else {}

	def register(self):
		payload = {
			'token': self.token,
			'serverName': self.server_name,
			**self.collect_stats(),
		}
		self.post_json('/agent/register', payload)

	def heartbeat(self):
		payload = {
			'token': self.token,
			'serverName': self.server_name,
			**self.collect_stats(),
		}
		self.post_json('/agent/heartbeat', payload)

	def run(self):
		print('[Agent] HTTP Agent started')
		while True:
			try:
				self.register()
				print('[Agent] registered')
				break
			except Exception as e:
				print(f'[Agent] register failed: {e}')
				time.sleep(5)

		while True:
			try:
				self.heartbeat()
				time.sleep(2)
			except Exception as e:
				print(f'[Agent] heartbeat failed: {e}')
				time.sleep(5)


if __name__ == '__main__':
	config_path = os.getenv('CONFIG_PATH', '/etc/spark-lab-agent/config.json')
	if not os.path.exists(config_path):
		print(f'Config not found: {config_path}')
		sys.exit(1)
	agent = SparkLabHTTPAgent(config_path)
	agent.run()
`

	c.Header("Content-Type", "text/x-python; charset=utf-8")
	c.String(http.StatusOK, script)
}

func (h *Handler) AgentPythonFile(c *gin.Context) {
	paths := []string{
		"../agent-python/agent.py",
		"./agent-python/agent.py",
	}

	var content []byte
	var err error
	for _, p := range paths {
		content, err = os.ReadFile(p)
		if err == nil {
			c.Header("Content-Type", "text/x-python; charset=utf-8")
			c.String(http.StatusOK, string(content))
			return
		}
	}

	c.JSON(http.StatusNotFound, gin.H{"message": fmt.Sprintf("agent.py not found: %v", err)})
}
